doc/designs/ech-api.md

```diff
@@ -0,0 +1,590 @@
+expose which web site was being accessed. Similarly, if the TLS server
+certificate for one web site were significantly larger or smaller than others,
+message sizes could reveal which web site was being visited.  For these
+reasons, we provide a way to enable additional ECH-specific padding of the
+Certificate, CertificateVerify and EncryptedExtensions messages sent from the
+server to the client during the handshake.
+
+To enable ECH-specific padding, one makes a call to:
+
+```c
+    SSL_CTX_set_options(ctx, SSL_OP_ECH_SPECIFIC_PADDING);
+```
+
+The default padding scheme is to ensure the following sizes for the plaintext
+form of these messages:
+
+| Message             | Minimum Size | Size is multiple of |
+| ------------------- | ------------ | ------------------- |
+| Certificate         | 1792         | 128                 |
+| CertificateVerify   | 480          | 16                  |
+| EncryptedExtensions | 32           | 16                  |
+
+The ciphertext form of these messages, as seen on the network in the record
+layer protocol, will usually be 16 octets more, due to the AEAD tag that is
+added as part of encryption.
+
+If a server wishes to have finer-grained control of these sizes, then it can
+make use of the ``SSL_CTX_ech_set_pad_sizes()`` or ``SSL_ech_set_pad_sizes()``
+APIs. Both involve populating an ``OSSL_ECH_PAD_SIZES`` data structure as
+described below in the obvious manner.
+
+```c
+/*
+ * Fine-grained ECH-spacific padding controls for a server
+ */
+typedef struct ossl_ech_pad_sizes_st {
+    size_t cert_min; /* minimum size */
+    size_t cert_unit; /* size will be multiple of */
+    size_t certver_min; /* minimum size */
+    size_t certver_unit; /* size will be multiple of */
+    size_t ee_min; /* minimum size */
+    size_t ee_unit; /* size will be multiple of */
+} OSSL_ECH_PAD_SIZES;
+
+int SSL_CTX_ech_set_pad_sizes(SSL_CTX *ctx, OSSL_ECH_PAD_SIZES *sizes);
+int SSL_ech_set_pad_sizes(SSL *s, OSSL_ECH_PAD_SIZES *sizes);
+```
+
+Client-side APIs
+----------------
+
+ECHConfig values contain a version, algorithm parameters, the public key to use
+for HPKE encryption and the ``public_name`` that is by default used for the
+outer SNI when ECH is attempted.
+
+Clients need to provide one or more ECHConfig values in order to enable ECH for
+an SSL connection. ``SSL_ech_set1_echconfig()`` and
+``SSL_CTX_set1_echconfig()`` allow clients to provide these to the library in
+binary, ascii-hex or base64 encoded format. Multiple calls to these functions
+will accumulate the set of ECHConfig values available for a connection. If the
+input value provided contains no suitable ECHConfig values (e.g. if it only
+contains ECHConfig versions that are not supported), then these functions will
+fail and return zero.
+
+```c
+int SSL_ech_set1_echconfig(SSL *s, const unsigned char *val, size_t len);
+int SSL_CTX_ech_set1_echconfig(SSL_CTX *ctx, const unsigned char *val,
+                               size_t len);
```

Comment:
Mostly because that's how they arrive at the client application. With the current APIs there is no need for a TLS client to understand what's inside an ECHConfigList and nor should there be.  E.g. for curl, the application layer code never needs to look inside the ECHConfigList at all.

For some "advanced" uses of HTTPS RRs that contain ECHConfigList values, a client might want to compare the public name or algs inside an ECHConfigList against other SvcParamKey fields of the HTTPS/SVCB RR, but my guess is that'll be uncommon.

So for clients, I'm fairly sure a binary buffer is correct.

Reply:
Separating out into a different object has other advantages:

- It separates the parsing of the binary/ascii-hex/base-64 format, from its use (so if you reuse it you don't have to parse it every time)
- It provides type safety so it is clear exactly what kind of object we expect to be passed, and whether it is an ECHConfig value, or and ECHConfigList, or something else
- It provides extensibility
- It means we don't have to pass around the pointer and its length everywhere

Like I said, I really dislike the idea of passing binary buffers around.